Extract Symolic Links Correctly

blender opened the pull request:
- [X] BugFix: When archiving, `FileMode` must be preserved along with adding symlink mode
- [X] Feature: When extracting, preserve symlinks as such if specified 
- [x] Tests: Added test for extraction
- [x] Tests: Address #38 


@jgm about task 2 maybe symlinks should be preserved always when extracting respecting the archive format

jgm commented:
> @jgm about task 2 maybe symlinks should be preserved always when extracting respecting the archive format

Yes, that would be okay with me.

blender commented:
@jgm All right, I will make that happen then. 

Also, I'm not sure why but the build randomly fails on Windows because of [cassava](https://hackage.haskell.org/package/cassava).

I noticed that CI for macOS is gone now. Is that intentional?

The problem with the tests was that the target of the symlink was not correctly specified. Path must be either absolute or relative to the link itself.

jgm commented:
+++ Tommaso Piazza [Jan 21 18 16:23 ]:
>   [1]@jgm All right, I will make that happen then.
>
>   Also, I'm not sure why but the build randomly fails on Windows because
>   of [2]cassava.

That's because of a bug in the cached version of stack. I've
deleted the cache and will run the build again.

>   I noticed that CI for macOS is gone now. Is that intentional?

No, I just copied the travis script from another project.
But I've stopped doing the CI builds for macOS in general
(partly because I develop on macOS), because they weren't
giving me much useful information and they tend to be slow.

>   The problem with the tests was that the target of the symlink was not
>   correctly specified. Path must be either absolute or relative to the
>   link itself.

Aha!

jgm commented:
Can you fix code to avoid compiler warnings, except those that are unavoidable due to changes in base? (e.g. warnings about importing Control.Applicative in later ghc versions where prelude supplies the needed function)

blender commented:
@jgm about preserving symlinlikns by default I'm still all for it but: I think you should know that this is a behavior change (potentially breaking for someone) that people might not expect since the library has never supported symlinks before.

Leaving preservation of symbolic links as an option during extraction will not make this change breaking while giving whoever is interested the option to do so.

now, to the code:

- removed warnings
- added full roundtrip test for symlinks
- maybe add a test where the lib extracts an archive made with `zip` ?
- maybe add a test where unzip extract an archive made by the lib?

jgm commented:
Re the travis failure: looks like you need to import Control.Applicative in the test suite.

No need to run hlint or stylish haskell; I don't think I've ever done that on this code.  WE might want to think about doing it later, though.

Re the extra tests:  we have a test now that reads an external zip; I just include the zip in extra-source-files in zip-archive.cabal.  So it would be easy to add another test like that that includes a symlink.  If you have time, I think this would be a good idea, but it's not essential.

blender commented on src/Codec/Archive/Zip.hs:
```diff
@@ -342,13 +352,50 @@ writeEntry opts entry = do
+-- function behaves like `writeEntry`.
+writeSymbolicLinkEntry :: [ZipOption] -> Entry -> IO ()
+writeSymbolicLinkEntry opts entry = do
+  if OptPreserveSymbolicLinks `notElem` opts
```
Should it do this by default without checking for OptPreserveSymbolicLinks while writing? If so, this potentially is a breaking behavior change.

blender commented:
@jgm 

- Added test for compatibility with `zip`, `unzip` CLI
- Re added the import (this produces a warning on recent versions of GHC)
- Need decision for this https://github.com/jgm/zip-archive/pull/39/files#r162938605

and then we're done :)

jgm commented:
Let's avoid the breaking behavior change.
Looks like this is good to go, then!
I'll merge, many thanks.
